compress.cc

```diff
@@ -15,10 +15,131 @@
 #include "utils/reusable_buffer.hh"
 #include "sstables/compress.hh"
 #include "sstables/exceptions.hh"
+#include "utils/hashers.hh"
 
 #include "compress.hh"
 #include "exceptions/exceptions.hh"
 #include "utils/class_registrator.hh"
+#include "sstables/sstable_compressor_factory.hh"
+#include "gms/feature_service.hh"
+#include "seastar/core/metrics.hh"
+#include "seastar/core/sharded.hh"
+
+// SHA256
+using dict_id = std::array<std::byte, 32>;
+class sstable_compressor_factory_impl;
+
+static seastar::logger compressor_factory_logger("compressor_factory");
+
+// Holds a raw dictionary blob (without algorithm-specific hash tables).
+// raw dicts might be used (and kept alive) directly by compressors (in particular, lz4 decompressor)
+// or referenced by algorithm-specific dicts.
+class raw_dict : public enable_lw_shared_from_this<raw_dict> {
+    mutable sstable_compressor_factory_impl* _owner;
+    dict_id _id;
+    std::vector<std::byte> _dict;
+public:
+    raw_dict(sstable_compressor_factory_impl& owner, dict_id key, std::span<const std::byte> dict);
+    ~raw_dict();
+    const std::span<const std::byte> raw() const {
+        return _dict;
+    }
+    dict_id id() const {
+        return _id;
+    }
+    void disown() const noexcept;
+};
+
+// A custom allocator for zstd, so that we can track its memory usage.
+struct zstd_callback_allocator {
+    using callback_type = std::function<void(ssize_t)>;
+    callback_type _callback;
+    using self = zstd_callback_allocator;
+    zstd_callback_allocator(callback_type cb) : _callback(std::move(cb)) {}
+    zstd_callback_allocator(self&&) = delete;
+    ZSTD_customMem as_zstd_custommem() & {
+        return ZSTD_customMem{
+            .customAlloc = [] (void* opaque, size_t n) -> void* {
+                auto addr = malloc(n);
```

Comment:
What if it returns null?

Reply:
When the allocator returns nullptr to zstd, zstd is supposed to fail at whatever it was doing and return an error code indicating an allocation failure.

(And if you are asking about `malloc_usable_size`, it's supposed to return 0 when called on a nullptr).

Reply:
> and return an error code indicating an allocation failure.

And we check the error code and throw an exception.

Reply:
Yes malloc_usable_size. I expect zstd to handle malloc errors. If malloc_usable_size is prepared, all is well.